### PR TITLE
RHAIENG-3062: tests(jupyter/trusty/test_notebook.ipynb‎): add comprehensive e2e tests for TrustyAI

### PR DESCRIPTION
**Pull request by ayush17**
## Summary
Expands the TrustyAI test notebook with comprehensive tests for packages declared in `pyproject.toml` that were not previously covered.

### PyTorch & Deep Learning
- **torch**: tensor operations, CUDA availability check
- **transformers**: Hugging Face pipeline and AutoModel imports
- **datasets**: Hugging Face datasets library
- **accelerate**: distributed training support

### ONNX Conversion
- **skl2onnx**: sklearn model to ONNX conversion (with actual model conversion test)
- **onnxconverter-common**: ONNX utilities

### MLOps
- **MLflow**: experiment tracking imports
- **KFP**: Kubeflow Pipelines SDK
- **Kubeflow Training**: distributed training SDK
- **CodeFlare SDK**: distributed computing

### Data Science
- **plotly**: interactive plotting (in addition to existing matplotlib)

### Day-to-Day Functionality Tests
- `pip list`: package enumeration
- `pip show trustyai`: package info retrieval
- `pip check`: dependency validation (ensures no broken deps)

## Files Changed
- `jupyter/trustyai/ubi9-python-3.12/test/test_notebook.ipynb`

## Test Plan
- [ ] Run test notebook in TrustyAI container
- [ ] Verify all imports succeed
- [ ] Verify all test cases pass

## References
- Packages from: `jupyter/trustyai/ubi9-python-3.12/pyproject.toml`
- Datascience deps from: `dependencies/odh-notebooks-meta-workbench-datascience-deps/pyproject.toml`

[RHAIENG-3062](https://redhat.atlassian.net/browse/RHAIENG-3062)

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **Tests**
  * Notebook test suite now skips unsupported CPU architectures, adds checks for core ML libraries and basic tensor behavior, tightens dependency version validation, simplifies pass logs, and fails CI if tests are unsuccessful.

* **Chores**
  * Added Feast to project dependencies (duplicate entry introduced) and refreshed many pinned package versions and lockfiles.

* **Manifests**
  * Bumped annotated dependency versions (notably Boto3 and MySQL connector) in image metadata.

### CODE REVIEW

**Comment by jiridanek** [comment]
@coderabbitai /lgtm, but, the feast tests don't pass in gha; @ayush17 should feast be installed on trustyai?

File: jupyter/trustyai/ubi9-python-3.12/test/test_notebook.ipynb [hunk 1]
```diff
@@ -1,131 +1,370 @@
...
+        "    def test_sklearn_model_training(self):\n",
+        "        \"\"\"Test scikit-learn model training.\"\"\"\n",
+        "        X, y = datasets.make_classification(n_samples=100, n_features=4, random_state=42)\n",
+        "        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)\n",
+        "        from sklearn.ensemble import RandomForestClassifier\n",
+        "        clf = RandomForestClassifier(n_estimators=10, random_state=42)\n",
+        "        clf.fit(X_train, y_train)\n",
+        "        score = clf.score(X_test, y_test)\n",
+        "        self.assertGreater(score, 0.5)\n",
+        "        print(f\"Scikit-learn version: {sklearn.__version__}\")\n",
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply jupyter/trustyai/ubi9-python-3.12/test/test_notebook.ipynb]: @coderabbitai is right, @ayush17 consider making more deterministic like that

**Comment by atheo89** [comment]
>should feast be installed on trustyai?

I would vote yes, as this missing package was probably overlooked and not included initially. I also recall something about adding it to images derived from datascience.

**Review by atheo89 — APPROVED** [review]
/lgtm

**Review by jiridanek — APPROVED** [review]
conditional /approve, @ayush17 pls take look at rabbit comments and fix the relevant ones, for example https://github.com/opendatahub-io/notebooks/pull/3363#discussion_r3072268595 is imo a good improvement, one line change

**Comment by ysok** [comment]
Please note on the failed tests:

image

File: jupyter/trustyai/ubi9-python-3.12/test/test_notebook.ipynb [hunk 2]
```diff
@@ -1,131 +1,398 @@
...
+        "        import datasets\n",
+        "        print(f\"Datasets version: {datasets.__version__}\")\n",
+        "\n",
+        "    def test_accelerate_import(self):\n",
+        "        \"\"\"Test accelerate library for distributed training.\"\"\"\n",
+        "        self.assertTrue(hasattr(accelerate, 'Accelerator'))\n",
+        "        print(f\"Accelerate version: {accelerate.__version__}\")\n",
+        "\n",
+        "\n",
+        "class TestMLOpsPackages(unittest.TestCase):\n",
```
Comment (atheo89) [inline jupyter/trustyai/ubi9-python-3.12/test/test_notebook.ipynb]: We don't need this tests here because he have them already covered by the datascience notebook https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi9-python-3.12/test/test_notebook.ipynb

**Review by atheo89 — CHANGES_REQUESTED** [review]
Hey Ayush! We don’t need to reintroduce tests that are already covered in the datascience/test_notebook.ipynb. Just focus on the packages listed under TrystyAI’s pyproject.toml, specifically: "transformers", "datasets", and "accelerate" and "torch", the rest are already covered

**Comment by atheo89** [comment]
@ayush17 any updates on this?

**Comment by atheo89** [comment]
Could you take a look at fixing the tests?

To speed things up and avoid waiting for the GHA execution, you can spin up a TrustyAI notebook in a cluster and run the test directly there. Make sure to include the expected_versions.json file as well.

Below is the AI-generated explanation from Copilot (with my comment on the Feast case):

My comment for Feast: Theoretically, it shouldn’t be None because we now install it in the TrustyAI notebook. If it is None, then either something is wrong with the installation or the test case itself is problematic.

The job is failing due to **two test failures** in the datascience ubi9 notebook:

1. **`TestFeast.test_version`** - `TypeError: expected string or bytes-like object, got 'NoneType'`
2. **`TestPyMongo.test_version`** - Version comparison failure

## Root Cause

The tests are attempting to parse version strings, but one or both libraries are returning `None` instead of a proper version string. This typically happens when:
- The package isn't installed correctly
- The version attribute is missing or malformed
- A regex pattern is trying to match against `None`

## Solution

You need to fix the test code to handle `None` values gracefully. Based on the error pattern, here's the recommended fix:

**For `TestFeast.test_version`:**
```python
def test_version(self):
    import feast
    version = feast.__version__
    # Add null check before regex operations
    assert version is not None, "Feast version is None"
    assert re.search(r'\d+\.\d+\.\d+', version), f"Invalid version format: {version}"
```

**For `TestPyMongo.test_version`:**
```python
def test_version(self):
    import pymongo
    version = pymongo.__version__
    # Add null check
    assert version is not None, "PyMongo version is None"
    assert isinstance(version, str), f"Expected string version, got {type(version)}"
    # Your version validation logic
```

## Additional Context

The test file is located in `tests/containers/workbenches/workbench_image_test.py`. You should:

1. **Add defensive null checks** before any string operations on version attributes
2. **Verify package installation** - Both `feast` and `pymongo` may not be properly installed in the container
3. **Check package specifications** - Ensure the container's dependency files include these packages with compatible versions

The security tests are also being skipped (due to [RHAIENG-509](https://redhat.atlassian.net/browse/RHAIENG-509)), which is a separate known issue not affecting this failure.

**Review by atheo89 — APPROVED** [review]
